### DISCUSSION TITLE
Saving to PC from Wasm (using `FileSavePicker`)

### DISCUSSION THREAD

**Discussion by BertrandDeSaintBezier** [unoplatform/uno#13857] [Q&A] [answered]
Trying to save `SKImage` from browser to disk. This works fine on Windows/WinUI, but in Wasm it seems to generate a corrupted file. 

I'm guessing that this is the "temp" file that is [referred to in the documentation](https://platform.uno/docs/articles/features/windows-storage-pickers.html#file-system-access-api-pickers), but I don't understand when or how the real data is written to the soruce file and the temp one discarded. I could be completely off-topic though.

```c#
public async void SaveQrCodeToLocation(string fileName)
{
    var fileSavePicker = new FileSavePicker();
    fileSavePicker.SuggestedStartLocation = PickerLocationId.PicturesLibrary;
    fileSavePicker.CommitButtonText = "Download";
    fileSavePicker.SuggestedFileName = $"{fileName}.png";
    fileSavePicker.FileTypeChoices.Add("Save for Print", new List<string>() { ".png", ".jpg" });
    fileSavePicker.FileTypeChoices.Add("Save for Web", new List<string>() { ".webp" });

    var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(App.AppWindow);
    WinRT.Interop.InitializeWithWindow.Initialize(fileSavePicker, hwnd);

    StorageFile savedFile = await fileSavePicker.PickSaveFileAsync();
    if (savedFile != null)
    {
        CachedFileManager.DeferUpdates(savedFile);

        var encoder = savedFile.FileType switch
        {
            ".png" => SKEncodedImageFormat.Png,
            ".jpeg" => SKEncodedImageFormat.Jpeg,
            ".webp" => SKEncodedImageFormat.Webp,
            _ => SKEncodedImageFormat.Png
        };

        using (var data = _skImage?.Encode(encoder, 100))
        using (var stream = File.OpenWrite(savedFile.Path))
        {
            data?.SaveTo(stream);
        }

        await CachedFileManager.CompleteUpdatesAsync(savedFile)
    }
    else
    {
        // No folder was picked or the dialog was cancelled.
    }
}
```

**Comment by MartinZikmund** [comment]
@BertrandDeSaintBezier The underlying JS File API is asynchronous, so it is possible that the reason for the corrupted file is that you are using synchronous methods instead - is there a `SaveToAsync` method that you could `await` instead? And could you try adding a `FlushAsync` inside the `using` block? The documentation is not too clear here, I need to add this information

**Reply by BertrandDeSaintBezier** [reply]
As suggested, I tried doing this with an async method instead of the synchronous `SaveTo()` that I was using. Unfortunately, this works identically on Windows and Wasm, respectively - that is to say that on Windows, the file is saved to disk and opens just fine, and on Wasm, I get the same "corrupted" file behaviour.
```c#
using (SKData? skData = _qrCodeImage?.Encode(encoder, 100))
{
    if (skData != null)
    {
        var data = skData.ToArray();
        using (FileStream fs = new FileStream(savedFile.Path, FileMode.Create, FileAccess.Write))
        {
            await fs.WriteAsync(data, 0, data.Length);
        }
    }
}
```
I also tried this using a `Flush()` and `FlushAsync()`, but no such luck 🤔.

**Reply by jeromelaban** [reply]
Thanks for the update. Could you open an issue [with a repro](https://github.com/unoplatform/uno/blob/e85bf7a378aa0041aeb0a8d6b8536b21fb9be4c3/doc/articles/uno-howto-create-a-repro.md)?

**Reply by BertrandDeSaintBezier** [reply]
Just opened [an issue](https://github.com/unoplatform/uno/issues/13892#issue-1925980911) for this.

**Reply by BertrandDeSaintBezier** [reply, marked as answer]
Marking this as answered for now, found a workaround in https://github.com/unoplatform/uno/issues/13892#issuecomment-1757397309
